### ISSUE TITLE
Inviting a user to an E2EE room does not share keys for history with them, causing UISIs everywhere.

### ISSUE THREAD

**Issue by ara4n** [element-hq/element-x-ios#1496]
### Steps to reproduce

1. In EX, Invite a user to an E2EE room.
2. The user accepts the invite
3. They never receive the keys for the shared history, so will not be able to catch up on history.
4. EX will show these messages as UISIs rather than surpressing them.

#### What did you expect?

If you invite a user to a room, you should share them the keys they need to decrypt the messages they have permission to. (RHUL might have undermined this, given it lets malicious servers fake invites to steal keys, in which case we might instead need to wait until we have client-controlled group membership).

#### What happened instead?

UISIs everywhere.

### Application version

343

### Will you send logs?

No

**Comment by kegsay** [comment]
I don't think this is true anymore. The complement-crypto tests at least pass for invited users.

**Comment by stefanceriu** [comment]
I think K is right, closing

**Comment by ara4n** [comment]
Sorry, it looks like the bug wasn't clear enough - this bug is definitely still open. When you invite a user into a room, EX does *not* share the historical keys with the user.

This is:
 * https://github.com/matrix-org/matrix-rust-sdk/issues/580 at the rust level
 * https://github.com/matrix-org/matrix-rust-sdk/pull/2650 as an implementation at rust

...but is snarled in RHUL fallout still. But from a product perspective, it's a real black eye.

**Comment by kegsay** [comment]
Ah, it would help if you didn't mention invites then. This is a general "we don't share historical keys" bug, invites are not a pre-req.

**Comment by ara4n** [comment]
but this is specifically about invites! the missing behaviour is that when Alice invites Bob to a room with shared hist viz, she should (in theory) use MSC3061 to send a tonne of keyshares for the history in that room so that Bob can actually read history.

In other words, it's the EX implementation of:
 * https://github.com/matrix-org/matrix-js-sdk/pull/1640
 * https://github.com/element-hq/element-ios/issues/4947
 * https://github.com/element-hq/element-android/issues/4153

Now, i think these got backed out post-RHUL, which is why this is now all in limbo, but from a product perspective i'm trying to point out that it's an awful experience and we've regressed here without a clear path forwards.

**Comment by murlock1000** [comment]
The recent migration of element-web to the new rust crypto system without supporting [MSC3061](https://github.com/matrix-org/matrix-spec-proposals/pull/3061) has broken some of systems that relied on this feature. Specifically, the functionality where a bot would invite whitelisted users to a room and share room keys with them so they could see historic messages.

What is the current solution for an automated system that invites users to a room with existing messages and enables them to see historic messages now?

**Comment by andybalaam** [comment]
This is now fixed via https://github.com/element-hq/element-meta/issues/2829 !  (Note: only when a user is explicitly invited.)
